xtask/src/swift.rs

```diff
@@ -0,0 +1,85 @@
+use std::fs;
+
+use clap::{Args, Subcommand};
+use xshell::{cmd, pushd};
+
+use crate::{workspace, Result};
+
+#[derive(Args)]
+pub struct SwiftArgs {
+    #[clap(subcommand)]
+    cmd: Option<SwiftCommand>,
+}
+
+#[derive(Subcommand)]
+enum SwiftCommand {
+    /// Build as a static lib.
+    BuildLibrary,
+    /// Build as an XCFramework.
+    BuildFramework,
+}
+
+impl SwiftArgs {
+    pub fn run(self) -> Result<()> {
+        let _p = pushd(&workspace::root_path()?)?;
+
+        match self.cmd {
+            Some(cmd) => match cmd {
+                SwiftCommand::BuildLibrary => build_library(),
+                SwiftCommand::BuildFramework => build_xcframework(),
+            },
+            None => build_library(),
+        }
+    }
+}
+
+fn build_library() -> Result<()> {
+    println!("Running debug library build.");
+
+    let root_directory = workspace::root_path()?;
+    let target_directory = root_directory.join("target");
+    let ffi_directory = root_directory.join("bindings/apple/generated/matrix_sdk_ffi");
+    let swift_directory = root_directory.join("bindings/apple/generated/swift");
+
+    let release_type = "debug";
+    let static_lib_filename = "libmatrix_sdk_ffi.a";
+
+    fs::create_dir_all(ffi_directory.as_path())?;
+    fs::create_dir_all(swift_directory.as_path())?;
+
+    cmd!("cargo build -p matrix-sdk-ffi").run()?;
+
+    fs::rename(
+        target_directory.join(release_type).join(static_lib_filename),
+        ffi_directory.join(static_lib_filename),
+    )?;
+
+    cmd!(
+        "uniffi-bindgen generate
+    			--language swift
+    			--lib-file {ffi_directory}/{static_lib_filename}
+    			--out-dir {ffi_directory}
+				{root_directory}/bindings/matrix-sdk-ffi/src/api.udl"
+    )
+    .run()?;
```

Comment:
maybe that is for a future PR, but I'd really prefer if we could switch to using [`uniffi_bindgen::swift::generate_bindings` library function](https://docs.rs/uniffi_bindgen/latest/uniffi_bindgen/bindings/swift/gen_swift/fn.generate_bindings.html) here rather than run the system-wide installed command. This would also allow us to ensure that the exact right version is used across branches by pinning the dependency in the `Cargo.toml`.

Reply:
I think that's not a use case properly supported by upstream, but it could work. Still might be worth dicussing with the UniFFI people on Matrix first. The docs of https://docs.rs/uniffi_bindgen/latest/uniffi_bindgen/bindings/fn.compile_bindings.html hint at the programmatic API only serving testing purposes on their side.

Reply:
wat? Oh, noes ... I really wanted to that ... switching an globally installed binary upon branch-switching is so annoying...

Reply:
It could very well still work, the binary is a very small shim over the library. I just wanted to note it's not *officially* supported (AFAIU).